tsdb/store.go

```diff
@@ -478,21 +471,80 @@ func (s *Store) DeleteMeasurement(database, name string) error {
 		return influxql.ErrMeasurementNotFound(name)
 	}
 
+	seriesKeys := m.SeriesKeys()
+
+	s.mu.RLock()
+	shards := s.filterShards(func(sh *Shard) bool {
+		return sh.database == database
+	})
+	s.mu.RUnlock()
+
+	if err := s.walkShards(shards, func(sh *Shard) error {
+		if err := sh.DeleteMeasurement(m.Name, seriesKeys); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
 	// Remove measurement from index.
 	db.DropMeasurement(m.Name)
 
-	// Remove underlying data.
+	return nil
+}
+
+// filterShards returns a slice of shards where fn returns true
+// for the shard.
+func (s *Store) filterShards(fn func(sh *Shard) bool) []*Shard {
+	shards := make([]*Shard, 0, len(s.shards))
 	for _, sh := range s.shards {
-		if sh.database != database {
-			continue
+		if fn(sh) {
+			shards = append(shards, sh)
 		}
+	}
+	return shards
+}
 
-		if err := sh.DeleteMeasurement(m.Name, m.SeriesKeys()); err != nil {
-			return err
-		}
+// walkShards apply a function to each shard in parallel.  If any of the
+// functions return an error, the first error is returned.
+func (s *Store) walkShards(shards []*Shard, fn func(sh *Shard) error) error {
+	// struct to hold the result of opening each reader in a goroutine
+	type res struct {
+		s   *Shard
+		err error
 	}
 
-	return nil
+	throttle := newthrottle(runtime.GOMAXPROCS(0))
```

Comment:
There's a small chance, but someone could manually override `GOMAXPROCS` to a smaller value than `NumCPU`, in which case we would have increased contention for resources. Reading the current value of `GOMAXPROCS` is probably the better solution, especially since Go 1.5, where it defaults to the number of available CPU cores.